### PR TITLE
syspolicy: add exit node related policies

### PR DESCRIPTION
**Pull request by clairew**
Adds policy keys ExitNodeID and ExitNodeIP.
Uses the policy keys to determine the exit node in preferences. Fixes tailscale/corp#15683
Also makes syspolicy handler public for mocking in tests

### CODE REVIEW

File: ipn/ipnlocal/local.go [hunk 1]
```diff
@@ -1280,6 +1281,28 @@ func (b *LocalBackend) updateNetmapDeltaLocked(muts []netmap.NodeMutation) (hand
...
+			return false
+		}
+		prefs.ExitNodeID = tailcfg.StableNodeID(exitNodeID)
+		prefs.ExitNodeIP = netip.Addr{}
+		return true
+	}
+
+	if exitNodeIPStr, _ := syspolicy.GetString(syspolicy.ExitNodeIP, ""); exitNodeIPStr != "" {
+		exitNodeIP, err := netip.ParseAddr(exitNodeIPStr)
+		if !exitNodeIP.IsValid() && err == nil {
```
Comment (agottardo) [inline ipn/ipnlocal/local.go:1296]: Should we log something if it's not valid (err != nil)?

File: ipn/ipnlocal/local_test.go [hunk 2]
```diff
@@ -1333,3 +1334,118 @@ func (rc *routeCollector) AdvertiseRoute(pfx netip.Prefix) error {
...
+				key: test.key,
+				s:   val,
+			})
+			pm := must.Get(newProfileManager(new(mem.Store), t.Logf))
+			pm.prefs = test.prefs.View()
+			must.Do(pm.SetPrefs(pm.prefs.AsStruct().View(), ""))
+			b.pm = pm
+			setExitNodeID(b.pm.prefs.AsStruct(), new(netmap.NetworkMap))
+			b.SetPrefs(pm.CurrentPrefs().AsStruct())
+			//	fmt.Printf("asdfsdf %v %v", test.key, test.name)
```
Comment (agottardo) [inline ipn/ipnlocal/local_test.go:1434]: Cleanup?

File: ipn/ipnlocal/local_test.go [hunk 3]
```diff
@@ -1333,3 +1334,118 @@ func (rc *routeCollector) AdvertiseRoute(pfx netip.Prefix) error {
...
+			switch test.key {
+			case syspolicy.ExitNodeID:
+				got := b.pm.prefs.ExitNodeID()
+				if got != tailcfg.StableNodeID(test.want) {
+					t.Errorf("got %v want %v", got, test.want)
+				}
+			case syspolicy.ExitNodeIP:
+				got := b.pm.prefs.ExitNodeIP()
+				fmt.Printf("helllo %v %v %v", test.name, got, val)
+				//		fmt.Printf("asdfss %v %v", test.want, netip.MustParseAddr(test.want))
```
Comment (agottardo) [inline ipn/ipnlocal/local_test.go:1444]: same

**Review by agottardo — APPROVED** [review]
LGTM!

File: ipn/ipnlocal/local.go [hunk 4]
```diff
@@ -1288,6 +1289,28 @@ func (b *LocalBackend) updateNetmapDeltaLocked(muts []netmap.NodeMutation) (hand
...
+			if prefs.ExitNodeID == "" && prefs.ExitNodeIP == exitNodeIP {
+				return false
+			}
```
Comment (sailorfrag) [inline ipn/ipnlocal/local.go:1307]: This case shouldn't return, it needs to be converted into an `ExitNodeID` later in this func.

File: ipn/ipnlocal/local.go [hunk 5]
```diff
@@ -1300,6 +1301,29 @@ func (b *LocalBackend) updateNetmapDeltaLocked(muts []netmap.NodeMutation) (hand
...
+	if exitNodeIDStr, _ := syspolicy.GetString(syspolicy.ExitNodeID, ""); exitNodeIDStr != "" {
+		exitNodeID := tailcfg.StableNodeID(exitNodeIDStr)
+		if prefs.ExitNodeID != exitNodeID {
+			prefs.ExitNodeID = tailcfg.StableNodeID(exitNodeID)
+			prefs.ExitNodeIP = netip.Addr{}
+			return true
+		}
+	}
+
+	var oldExitNodeID tailcfg.StableNodeID
```
Comment (sailorfrag) [inline ipn/ipnlocal/local.go:1313]: While I think it's going to be harmless in the end to only set `oldExitNodeID` in the "syspolicy has an `ExitNodeIP` configured" path, it took a bunch of thinking it through to confirm that. I think it would be simpler just to always capture the old exit node ID.

```suggestion
	oldExitNodeID := prefs.ExitNodeID
```
and then don't bother setting it on line 1320

File: util/syspolicy/handler.go [hunk 6]
```diff
@@ -6,11 +6,12 @@ package syspolicy
...
 	"errors"
 	"sync/atomic"
+	"testing"
 )
 
 var (
-	handlerUsed atomic.Bool
-	handler     Handler = defaultHandler{}
+	PolicyHandlerUsed atomic.Bool
+	PolicyHandler     Handler = defaultHandler{}
```
Comment (sailorfrag) [inline util/syspolicy/handler.go:14]: Don't export these

File: ipn/ipnlocal/local_test.go [hunk 7]
```diff
@@ -1330,3 +1331,317 @@ func (rc *routeCollector) AdvertiseRoute(pfx netip.Prefix) error {
...
+	exitNodeIP string
+	bothKeys   bool
+	err        error
+}
+
+func (h *mockSyspolicyHandler) ReadString(key string) (string, error) {
+	if key != string(h.key) {
+		return "", h.err
+	}
+	if h.bothKeys { // when both keys are used
```
Comment (sailorfrag) [inline ipn/ipnlocal/local_test.go:1358]: This will never be reached when it matters, the key already won't match so it'll have returned `"", h.err` above.

It's probably easier to just have a pair of booleans that control if the ID can be looked up or the IP can be looked up. Either that or always return the keys as existing, and they just return `""` if unset for the test.

File: ipn/ipnlocal/local_test.go [hunk 8]
```diff
@@ -1330,3 +1331,317 @@ func (rc *routeCollector) AdvertiseRoute(pfx netip.Prefix) error {
...
+		name           string
+		key            syspolicy.Key
+		exitNodeID     string
+		exitNodeIP     string
+		prefs          *ipn.Prefs
+		exitNodeIPWant string
+		exitNodeIDWant string
+		prefsChanged   bool
+		nm             *netmap.NetworkMap
+		bothKeys       bool //indicates if we want to handle both ExitNodeID and ExitNodeIP keys in a test.
```
Comment (sailorfrag) [inline ipn/ipnlocal/local_test.go:1389]: The relationship between `key` and `bothKeys` is complicated and kind of redundant. I recommend instead having a pair of booleans to say if each one is in use. Or treat each one as in use if `exitNodeXX != ""`

Either way, I think getting rid of `key` and `bothKeys` is better because they're kind of confusing right now.

File: ipn/ipnlocal/local_test.go [hunk 9]
```diff
@@ -1330,3 +1331,283 @@ func (rc *routeCollector) AdvertiseRoute(pfx netip.Prefix) error {
...
+		nm             *netmap.NetworkMap
+	}{
+		{
+			name:           "ExitNodeID key is set",
+			exitNodeIDKey:  true,
+			exitNodeID:     "123",
+			prefs:          ipn.NewPrefs(),
+			exitNodeIDWant: "123",
+			prefsChanged:   true,
+			nm:             new(netmap.NetworkMap),
```
Comment (sailorfrag) [inline ipn/ipnlocal/local_test.go:1386]: optional: to reduce boilerplate when `nm` or `prefs` are empty, allow `nil` values (so they don't need to be mentioned at all in here) and then in the test loop, make an empty prefs or netmap if `tt.prefs == nil` or `tt.nm == nil` respectively

File: util/syspolicy/handler.go [hunk 10]
```diff
@@ -6,11 +6,12 @@ package syspolicy
 import (
 	"errors"
 	"sync/atomic"
+	"testing"
 )
 
 var (
-	handlerUsed atomic.Bool
-	handler     Handler = defaultHandler{}
+	policyHandlerUsed atomic.Bool
```
Comment (sailorfrag) [inline util/syspolicy/handler.go:13]: switch back to the original names? Two reasons:
1. The `policy` part is redundant considering this is the `syspolicy` package
2. Makes this PR smaller

**Review by sailorfrag — COMMENTED** [review]
Very close to finished!

File: ipn/ipnlocal/local.go [hunk 11]
```diff
@@ -1327,6 +1346,9 @@ func setExitNodeID(prefs *ipn.Prefs, nm *netmap.NetworkMap) (prefsChanged bool)
 			// reference it directly for next time.
 			prefs.ExitNodeID = peer.StableID()
 			prefs.ExitNodeIP = netip.Addr{}
+			if oldExitNodeID == prefs.ExitNodeID {
+				return false
+			}
 			return true
```
Comment (sailorfrag) [inline ipn/ipnlocal/local.go:1352]: This can be simplified down to

```suggestion
			return oldExitNodeID != prefs.ExitNodeID
```

File: ipn/ipnlocal/local_test.go [hunk 12]
```diff
@@ -1330,3 +1331,281 @@ func (rc *routeCollector) AdvertiseRoute(pfx netip.Prefix) error {
...
+	return "", h.err
+}
+
+func (h *mockSyspolicyHandler) ReadUInt64(key string) (uint64, error) {
+	return 0, h.err
+}
+
+func (h *mockSyspolicyHandler) ReadBoolean(key string) (bool, error) {
+	return false, h.err
+}
```
Comment (sailorfrag) [inline ipn/ipnlocal/local_test.go:1363]: I don't think `h.err` is ever being used, so let's trim it down to just the things that get used.

```suggestion
type mockSyspolicyHandler struct {
	t             *testing.T
	key           syspolicy.Key
	s             string
	exitNodeIDKey bool
	exitNodeIPKey bool
	exitNodeID    string
	exitNodeIP    string
}

func (h *mockSyspolicyHandler) ReadString(key string) (string, error) {
	if h.exitNodeIDKey && key == string(syspolicy.ExitNodeID) {
		return h.exitNodeID, nil
	}
	if h.exitNodeIPKey && key == string(syspolicy.ExitNodeIP) {
		return h.exitNodeIP, nil
	}
	return "", syspolicy.ErrNoSuchKey
}

func (h *mockSyspolicyHandler) ReadUInt64(key string) (uint64, error) {
	return 0, syspolicy.ErrNoSuchKey
}

func (h *mockSyspolicyHandler) ReadBoolean(key string) (bool, error) {
	return false, syspolicy.ErrNoSuchKey
}
```

File: ipn/ipnlocal/local_test.go [hunk 13]
```diff
@@ -1330,3 +1331,281 @@ func (rc *routeCollector) AdvertiseRoute(pfx netip.Prefix) error {
...
+				},
+			},
+		},
+		{
+			name:           "ExitNodeIP cleared if net map has corresponding node - policy matches prefs",
+			prefs:          &ipn.Prefs{ExitNodeIP: netip.MustParseAddr("127.0.0.1")},
+			exitNodeIDKey:  true,
+			exitNodeIPKey:  true,
+			exitNodeIP:     "127.0.0.1",
+			exitNodeID:     "123",
```
Comment (sailorfrag) [inline ipn/ipnlocal/local_test.go:1463]: I think the `ExitNodeID` is setting the pref directly rather than looking it up using the netmap.

```suggestion
			exitNodeIPKey:  true,
			exitNodeIP:     "127.0.0.1",
```

File: ipn/ipnlocal/local_test.go [hunk 14]
```diff
@@ -1330,3 +1331,281 @@ func (rc *routeCollector) AdvertiseRoute(pfx netip.Prefix) error {
...
+				},
+			},
+		},
+		{
+			name:           "ExitNodeIP cleared if net map has corresponding node - no policy set",
+			prefs:          &ipn.Prefs{ExitNodeIP: netip.MustParseAddr("127.0.0.1")},
+			exitNodeIDKey:  true,
+			exitNodeIPKey:  true,
+			exitNodeIP:     "",
+			exitNodeID:     "123",
```
Comment (sailorfrag) [inline ipn/ipnlocal/local_test.go:1499]: The description says no policy set but it's setting a policy

```suggestion
```

File: ipn/ipnlocal/local_test.go [hunk 15]
```diff
@@ -1330,3 +1331,281 @@ func (rc *routeCollector) AdvertiseRoute(pfx netip.Prefix) error {
...
+				},
+			},
+		},
+		{
+			name:           "ExitNodeIP cleared if net map has corresponding node - different exit node IP in policy",
+			exitNodeIDKey:  true,
+			exitNodeIPKey:  true,
+			prefs:          &ipn.Prefs{ExitNodeIP: netip.MustParseAddr("127.0.0.1")},
+			exitNodeIP:     "100.64.5.6",
+			exitNodeID:     "123",
```
Comment (sailorfrag) [inline ipn/ipnlocal/local_test.go:1535]: Again the `exitNodeID` is going to be the thing that actually makes the change, rather than the mechanism stated by the description.

```suggestion
			exitNodeIPKey:  true,
			prefs:          &ipn.Prefs{ExitNodeIP: netip.MustParseAddr("127.0.0.1")},
			exitNodeIP:     "100.64.5.6",
```

**Review by sailorfrag — COMMENTED** [review]
Just some test fixes remaining. I think the tests will still pass after the changes.

**Review by sailorfrag — APPROVED** [review]
